packages/component/src/BasicTranscript.js

```diff
@@ -61,31 +62,57 @@ function sameTimestampGroup(activityX, activityY, groupTimestamp) {
   return false;
 }
 
-const BasicTranscript = ({ activityRenderer, attachmentRenderer, className }) => {
+const BasicTranscript = ({ className }) => {
   const [{ activities: activitiesStyleSet, activity: activityStyleSet }] = useStyleSet();
   const [{ hideScrollToEndButton }] = useStyleOptions();
+  const [{ speechSynthesis, SpeechSynthesisUtterance } = {}] = useWebSpeechPonyfill();
   const [activities] = useActivities();
   const [groupTimestamp] = useGroupTimestamp();
-  const [{ speechSynthesis, SpeechSynthesisUtterance } = {}] = useWebSpeechPonyfill();
+  const renderAttachment = useRenderAttachment();
+  const renderActivity = useRenderActivity(renderAttachment);
 
   // We use 2-pass approach for rendering activities, for show/hide timestamp grouping.
   // Until the activity pass thru middleware, we never know if it is going to show up.
   // After we know which activities will show up, we can compute which activity will show timestamps.
   // If the activity does not render, it will not be spoken if text-to-speech is enabled.
-  const activityElements = activities.reduce((activityElements, activity) => {
-    const element = activityRenderer({
-      activity,
-      timestampClassName: 'transcript-timestamp'
-    })(({ attachment }) => attachmentRenderer({ activity, attachment }));
-
-    element &&
-      activityElements.push({
-        activity,
-        element
-      });
+  const activityElements = useMemo(
+    () =>
+      activities.reduce((activityElements, activity, index) => {
+        const element = renderActivity({
+          activity,
+          timestampClassName: 'transcript-timestamp'
+        });
+
+        element &&
+          activityElements.push({
+            activity,
+            element,
+            key: (activity.channelData && activity.channelData.clientActivityID) || activity.id || index,
+
+            // TODO: [P2] We should use core/definitions/speakingActivity for this predicate instead
+            shouldSpeak: activity.channelData && activity.channelData.speak
+          });
+
+        return activityElements;
+      }, []),
+    [activities, renderActivity]
+  );
 
-    return activityElements;
-  }, []);
+  const activityElements2 = useMemo(
```

Comment:
I don't get why we need `activityElements` and `activityElements2`

Reply:
`activityElements2` is the 2nd pass, which we make the decision whether a specific activity should be grouped with other activities in a timestamp group. We cannot do this in 1st pass due to lack of knowledge. The grouping decision is based on knowledge after the 1st pass is done.